core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -379,6 +379,41 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     }
   executorAllocationManager.foreach(_.start())
 
+  // Use reflection to instantiate listeners specified via the `spark.extraListeners` configuration
+  // or the SPARK_EXTRA_LISTENERS environment variable
+  try {
+    val listenerClassNames: Seq[String] = {
+      val fromSparkConf = conf.get("spark.extraListeners", "").split(',')
+      val fromEnvVar = Option(conf.getenv("SPARK_EXTRA_LISTENERS")).getOrElse("").split(',')
```

Comment:
Hm - I'm also skeptical about having an env var here. This same reasoning could be used to argue having an env var for every conf value right (which we definitely don't want). @JoshRosen at least for the use case I had in mind for this, we could on our own preprocess the conf files to deal with merging (i.e. no need to put extra logic in Spark).